src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -1820,6 +1907,253 @@ public string DeleteCounter(DocumentsOperationContext context, string documentId
+                        names.Modifications.Remove(item.Name);
+
+                        data.Modifications = new DynamicJsonValue(data);
+                        using (var old = data)
+                        {
+                            data = context.ReadObject(data, item.DocumentId, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
+                        }
+
+                        var newEtag = _documentsStorage.GenerateNextEtag();
+                        var newChangeVector = _documentsStorage.GetNewChangeVector(context, newEtag);
+
+                        using (Slice.From(context.Allocator, newChangeVector, out var cv))
+                        using (DocumentIdWorker.GetStringPreserveCase(context, collectionName.Name, out Slice collectionSlice))
+                        using (table.Allocate(out TableValueBuilder tvb))
+                        {
+                            tvb.Add(counterGroupKey);
+                            tvb.Add(Bits.SwapBytes(newEtag));
+                            tvb.Add(cv);
+                            tvb.Add(data.BasePointer, data.Size);
+                            tvb.Add(collectionSlice);
+                            tvb.Add(context.GetTransactionMarker());
+
+                            table.Set(tvb);
+                        }
+                    }
+
+                    try
+                    {
+                        var document = _documentsStorage.Get(context, item.DocumentId,
+                            throwOnConflict: true);
+
+                        var newData = document.Data.Clone(context);
+                        _documentDatabase.DocumentsStorage.Put(context, item.DocumentId, expectedChangeVector: null, newData, flags: document.Flags, nonPersistentFlags: document.NonPersistentFlags);
+                    }
+                    catch (Exception e)
+                    {
+                        throw new Exception($"Failed to get or update document '{item.DocumentId}'," +
+                                            $"Exception: {e.Message}");
+                    }
+                }
+            }
+
+            return deleted;
+        }
+
+        private long PurgeCounterTombstones(in long upto, DocumentsOperationContext context, CollectionName collectionName, long numberOfEntriesToDelete)
+        {
+            // delete counter tombstones entries
+
+            var counterTombstonesTableName = collectionName.GetTableName(CollectionTableType.CounterTombstones);
+            var tombstonesTable = context.Transaction.InnerTransaction.OpenTable(CounterTombstonesSchema, counterTombstonesTableName);
+            if (tombstonesTable == null || tombstonesTable.NumberOfEntries == 0 || numberOfEntriesToDelete <= 0)
+                return 0;
+
+            return tombstonesTable.DeleteBackwardFrom(CounterTombstonesSchema.FixedSizeIndexes[CollectionCounterTombstonesEtagsSlice], upto, numberOfEntriesToDelete);
```

Comment:
You are getting the counter tombstones from etag `0` and seek upwards, but delete backwards...
is that possible that you will delete something that wasn't handled yet?